{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f2842876-a817-42d9-9245-166ec78d348b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os \n",
    "sys.path.append(os.path.expanduser(f\"~/SSS_mount/insituCNV/InSituCNV\"))\n",
    "import insitucnv as icv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1d056b0-1300-4964-9a95-1bebff1f82c6",
   "metadata": {},
   "source": [
    "# Load AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "de01aff7-39ee-4f59-8940-03a37d9ffe45",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_path = \"/home/augusta/SSS_mount/insituCNV/data/simulated_CNV_data/lung_organoids_cnvclust.h5ad\"\n",
    "adata = sc.read_h5ad(adata_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f3638c5-956e-4e3e-9e99-08ec5e6240a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "N3_O1_AAACCCAAGCGTCAAG-1    secretory cell\n",
       "N3_O1_AAAGAACAGCGTATGG-1    secretory cell\n",
       "N3_O1_AAAGAACCACTATCGA-1    secretory cell\n",
       "N3_O1_AAAGGATTCGAGTCCG-1    secretory cell\n",
       "N3_O1_AAAGGGCGTTCTTAGG-1        basal cell\n",
       "                                 ...      \n",
       "N3_O2_GGTAGAGCAAGCACCC-1    secretory cell\n",
       "N3_O2_TTCAATCAGGGCAACT-1        basal cell\n",
       "N3_O2_TTGTGTTAGAACTGAT-1        basal cell\n",
       "N3_EX_CATGGTACAATTGCGT-1    secretory cell\n",
       "N3_EX_GGTAACTAGCGTGAGT-1    secretory cell\n",
       "Name: cell_type, Length: 1268, dtype: category\n",
       "Categories (3, object): ['ciliated cell', 'secretory cell', 'basal cell']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.cell_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a55d75ae-9be2-4145-ac69-025e506b855d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 ciliated cell\n",
      "1031 secretory cell\n",
      "234 basal cell\n"
     ]
    }
   ],
   "source": [
    "print(adata[adata.obs.cell_type=='ciliated cell'].n_obs, 'ciliated cell')\n",
    "print(adata[adata.obs.cell_type=='secretory cell'].n_obs, 'secretory cell')\n",
    "print(adata[adata.obs.cell_type=='basal cell'].n_obs, 'basal cell')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "389aa2c2-2f37-4173-baf3-2ba9b4f0a9b0",
   "metadata": {},
   "source": [
    "# Simulate the data to contain CNVs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "985811da-9f5c-4d80-a361-681ecd39447f",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### Step 1: Generate CNVs\n",
    "\n",
    "This function generates copy number variations (CNVs) based on a dictionary (**CNV_dict**) of genes (keys) and whether they should be **gain** or **loss** (value). \n",
    "\n",
    "Each CNV is centered around the gene and spans a **random size** between **min_size** and **max_size**.  \n",
    "The output is a **DataFrame** containing the CNV details:  \n",
    "\n",
    "- **Gene Name**\n",
    "- **Chromosome**\n",
    "- **CNV Size (bp)**\n",
    "- **CNV Type (gain/loss)**\n",
    "- **Start Position (bp)**\n",
    "- **End Position (bp)**\n",
    "\n",
    "**Parameters:**\n",
    "- **CNV_dict** *(dict)*: Dictionary mapping genes to CNV type ('gain' or 'loss').  \n",
    "- **min_size** *(int)*: Minimum CNV size in base pairs (bp).  \n",
    "- **max_size** *(int)*: Maximum CNV size in base pairs (bp).  \n",
    "- **gene_info_path** *(str)*: Path to a CSV file containing gene information.   \n",
    "  - Expected columns: `Gene name`, `Chromosome/scaffold name`, `Gene start (bp)`, `Gene end (bp)`.  \n",
    "- **save_csv** *(str, optional)*: If provided, saves the output DataFrame to this file.\n",
    "\n",
    "**Returns:**\n",
    "- **CNV_df** *(DataFrame)*: Contains the generated CNVs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "22663b9b-69c7-489f-b660-b851cb30e32a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbols</th>\n",
       "      <th>feature_is_filtered</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_reference</th>\n",
       "      <th>feature_biotype</th>\n",
       "      <th>feature_length</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>n_cells_by_counts</th>\n",
       "      <th>mean_counts</th>\n",
       "      <th>log1p_mean_counts</th>\n",
       "      <th>pct_dropout_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>629</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>486</td>\n",
       "      <td>0.008812</td>\n",
       "      <td>0.008774</td>\n",
       "      <td>94.534413</td>\n",
       "      <td>78.360051</td>\n",
       "      <td>4.373995</td>\n",
       "      <td>chr1</td>\n",
       "      <td>89295.0</td>\n",
       "      <td>133566.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000241860</th>\n",
       "      <td>AL627309.5</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000241860.7</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1025</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>581</td>\n",
       "      <td>0.010553</td>\n",
       "      <td>0.010498</td>\n",
       "      <td>93.466037</td>\n",
       "      <td>93.837299</td>\n",
       "      <td>4.552163</td>\n",
       "      <td>chr1</td>\n",
       "      <td>141474.0</td>\n",
       "      <td>173862.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000241599</th>\n",
       "      <td>AL627309.4</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000241599.1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>457</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>7</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>99.921278</td>\n",
       "      <td>2.190479</td>\n",
       "      <td>1.160171</td>\n",
       "      <td>chr1</td>\n",
       "      <td>160446.0</td>\n",
       "      <td>161525.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237491</th>\n",
       "      <td>LINC01409</td>\n",
       "      <td>False</td>\n",
       "      <td>LINC01409</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1059</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1591</td>\n",
       "      <td>0.035527</td>\n",
       "      <td>0.034910</td>\n",
       "      <td>82.107512</td>\n",
       "      <td>315.902356</td>\n",
       "      <td>5.758594</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778747.0</td>\n",
       "      <td>810065.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228794</th>\n",
       "      <td>LINC01128</td>\n",
       "      <td>False</td>\n",
       "      <td>LINC01128</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>1627</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2011</td>\n",
       "      <td>0.042585</td>\n",
       "      <td>0.041704</td>\n",
       "      <td>77.384166</td>\n",
       "      <td>378.669894</td>\n",
       "      <td>5.939302</td>\n",
       "      <td>chr1</td>\n",
       "      <td>825138.0</td>\n",
       "      <td>859446.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_symbols  feature_is_filtered       feature_name  \\\n",
       "gene_ids                                                               \n",
       "ENSG00000238009   AL627309.1                False  ENSG00000238009.6   \n",
       "ENSG00000241860   AL627309.5                False  ENSG00000241860.7   \n",
       "ENSG00000241599   AL627309.4                False  ENSG00000241599.1   \n",
       "ENSG00000237491    LINC01409                False          LINC01409   \n",
       "ENSG00000228794    LINC01128                False          LINC01128   \n",
       "\n",
       "                feature_reference feature_biotype feature_length feature_type  \\\n",
       "gene_ids                                                                        \n",
       "ENSG00000238009    NCBITaxon:9606            gene            629       lncRNA   \n",
       "ENSG00000241860    NCBITaxon:9606            gene           1025       lncRNA   \n",
       "ENSG00000241599    NCBITaxon:9606            gene            457       lncRNA   \n",
       "ENSG00000237491    NCBITaxon:9606            gene           1059       lncRNA   \n",
       "ENSG00000228794    NCBITaxon:9606            gene           1627       lncRNA   \n",
       "\n",
       "                 n_cells_by_counts  mean_counts  log1p_mean_counts  \\\n",
       "gene_ids                                                             \n",
       "ENSG00000238009                486     0.008812           0.008774   \n",
       "ENSG00000241860                581     0.010553           0.010498   \n",
       "ENSG00000241599                  7     0.000246           0.000246   \n",
       "ENSG00000237491               1591     0.035527           0.034910   \n",
       "ENSG00000228794               2011     0.042585           0.041704   \n",
       "\n",
       "                 pct_dropout_by_counts  total_counts  log1p_total_counts  \\\n",
       "gene_ids                                                                   \n",
       "ENSG00000238009              94.534413     78.360051            4.373995   \n",
       "ENSG00000241860              93.466037     93.837299            4.552163   \n",
       "ENSG00000241599              99.921278      2.190479            1.160171   \n",
       "ENSG00000237491              82.107512    315.902356            5.758594   \n",
       "ENSG00000228794              77.384166    378.669894            5.939302   \n",
       "\n",
       "                chromosome     start       end  \n",
       "gene_ids                                        \n",
       "ENSG00000238009       chr1   89295.0  133566.0  \n",
       "ENSG00000241860       chr1  141474.0  173862.0  \n",
       "ENSG00000241599       chr1  160446.0  161525.0  \n",
       "ENSG00000237491       chr1  778747.0  810065.0  \n",
       "ENSG00000228794       chr1  825138.0  859446.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7706856b-50d6-402d-9193-d524da74edb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "CNV_dict = { \n",
    "    'DIS3': 'loss',\n",
    "    'MECOM': 'loss',\n",
    "    'ERBB2': 'gain',\n",
    "    'CHD7': 'gain',\n",
    "    'HCK': 'gain',\n",
    "    'KEAP1': 'loss',\n",
    "    'MYD88': 'gain',\n",
    "    'TBX3': 'gain'\n",
    "}\n",
    "\n",
    "gene_file = (\"/home/augusta/SSS_mount/insituCNV/InSituCNV/Ensmbl_BioMart_gene_info.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1c18ec3-2127-465a-b4df-d8f178e65363",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CNV_df = icv.pp.generate_cnvs(CNV_dict, min_size=1000000, max_size=5000000, gene_info=gene_file, save_csv='CNVs_121224_simulation1.csv') \n",
    "# CNV_df = icv.pp.generate_cnvs(CNV_dict, min_size=5000000, max_size=10000000, gene_info=gene_file, save_csv='CNVs_310125_simulation2.csv') \n",
    "CNV_df = icv.pp.generate_cnvs(CNV_dict, min_size=10000000, max_size=20000000, gene_info=gene_file, save_csv='CNVs_310125_simulation3.csv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7fad1bc5-67d6-47b2-b4be-3fc1981eeedb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Size (bp)</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start (bp)</th>\n",
       "      <th>End (bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DIS3</td>\n",
       "      <td>13</td>\n",
       "      <td>19739671</td>\n",
       "      <td>loss</td>\n",
       "      <td>62897297</td>\n",
       "      <td>82636967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MECOM</td>\n",
       "      <td>3</td>\n",
       "      <td>11795339</td>\n",
       "      <td>loss</td>\n",
       "      <td>163475968</td>\n",
       "      <td>175271306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERBB2</td>\n",
       "      <td>17</td>\n",
       "      <td>14562701</td>\n",
       "      <td>gain</td>\n",
       "      <td>32427820</td>\n",
       "      <td>46990520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHD7</td>\n",
       "      <td>8</td>\n",
       "      <td>14661851</td>\n",
       "      <td>gain</td>\n",
       "      <td>53442459</td>\n",
       "      <td>68104309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCK</td>\n",
       "      <td>20</td>\n",
       "      <td>17670343</td>\n",
       "      <td>gain</td>\n",
       "      <td>23241855</td>\n",
       "      <td>40912197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>KEAP1</td>\n",
       "      <td>19</td>\n",
       "      <td>13300526</td>\n",
       "      <td>loss</td>\n",
       "      <td>3844578</td>\n",
       "      <td>17145104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MYD88</td>\n",
       "      <td>3</td>\n",
       "      <td>12730315</td>\n",
       "      <td>gain</td>\n",
       "      <td>31775631</td>\n",
       "      <td>44505945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TBX3</td>\n",
       "      <td>12</td>\n",
       "      <td>11391694</td>\n",
       "      <td>gain</td>\n",
       "      <td>108981368</td>\n",
       "      <td>120373062</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene name  Chromosome  Size (bp)  Type  Start (bp)   End (bp)\n",
       "0      DIS3          13   19739671  loss    62897297   82636967\n",
       "1     MECOM           3   11795339  loss   163475968  175271306\n",
       "2     ERBB2          17   14562701  gain    32427820   46990520\n",
       "3      CHD7           8   14661851  gain    53442459   68104309\n",
       "4       HCK          20   17670343  gain    23241855   40912197\n",
       "5     KEAP1          19   13300526  loss     3844578   17145104\n",
       "6     MYD88           3   12730315  gain    31775631   44505945\n",
       "7      TBX3          12   11391694  gain   108981368  120373062"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CNV_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "070c1828-a48f-402c-99de-262499e03d47",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Step 2: Create a CNV Template for AnnData\n",
    "This step maps the CNVs onto the genes stored in an **AnnData object (adata)** to create a CNV template.  \n",
    "\n",
    "The **CNV template** is a matrix that marks whether each gene is:  \n",
    "🔹 **Gained** *(+1)*  \n",
    "🔹 **Lost** *(-1)*  \n",
    "🔹 **Unchanged** *(0)*  \n",
    "\n",
    "**Process Overview**\n",
    "- Sort genes in **adata.var** by **chromosome** and **start position**.\n",
    "- Initialize an **empty CNV template matrix**.\n",
    "- Identify genes **within** CNV regions.\n",
    "- Map genes to their corresponding **CNV effect**.\n",
    "\n",
    "**Returns:**\n",
    "- A **CNV template DataFrame**, where rows correspond to CNVs and columns correspond to genes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec78a6a5-33ef-4eba-baa3-f855f7a756ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_ids</th>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <th>ENSG00000241860</th>\n",
       "      <th>ENSG00000241599</th>\n",
       "      <th>ENSG00000286448</th>\n",
       "      <th>ENSG00000237491</th>\n",
       "      <th>ENSG00000225880</th>\n",
       "      <th>ENSG00000228794</th>\n",
       "      <th>ENSG00000230368</th>\n",
       "      <th>ENSG00000272438</th>\n",
       "      <th>ENSG00000223764</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000287171</th>\n",
       "      <th>ENSG00000196664</th>\n",
       "      <th>ENSG00000228933</th>\n",
       "      <th>ENSG00000224294</th>\n",
       "      <th>ENSG00000122824</th>\n",
       "      <th>ENSG00000158639</th>\n",
       "      <th>ENSG00000269437</th>\n",
       "      <th>ENSG00000204025</th>\n",
       "      <th>ENSG00000165509</th>\n",
       "      <th>ENSG00000126895</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DIS3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MECOM</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERBB2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHD7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCK</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEAP1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYD88</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TBX3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 25691 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_ids  ENSG00000238009  ENSG00000241860  ENSG00000241599  ENSG00000286448  \\\n",
       "DIS3                  0.0              0.0              0.0              0.0   \n",
       "MECOM                 0.0              0.0              0.0              0.0   \n",
       "ERBB2                 0.0              0.0              0.0              0.0   \n",
       "CHD7                  0.0              0.0              0.0              0.0   \n",
       "HCK                   0.0              0.0              0.0              0.0   \n",
       "KEAP1                 0.0              0.0              0.0              0.0   \n",
       "MYD88                 0.0              0.0              0.0              0.0   \n",
       "TBX3                  0.0              0.0              0.0              0.0   \n",
       "\n",
       "gene_ids  ENSG00000237491  ENSG00000225880  ENSG00000228794  ENSG00000230368  \\\n",
       "DIS3                  0.0              0.0              0.0              0.0   \n",
       "MECOM                 0.0              0.0              0.0              0.0   \n",
       "ERBB2                 0.0              0.0              0.0              0.0   \n",
       "CHD7                  0.0              0.0              0.0              0.0   \n",
       "HCK                   0.0              0.0              0.0              0.0   \n",
       "KEAP1                 0.0              0.0              0.0              0.0   \n",
       "MYD88                 0.0              0.0              0.0              0.0   \n",
       "TBX3                  0.0              0.0              0.0              0.0   \n",
       "\n",
       "gene_ids  ENSG00000272438  ENSG00000223764  ...  ENSG00000287171  \\\n",
       "DIS3                  0.0              0.0  ...              0.0   \n",
       "MECOM                 0.0              0.0  ...              0.0   \n",
       "ERBB2                 0.0              0.0  ...              0.0   \n",
       "CHD7                  0.0              0.0  ...              0.0   \n",
       "HCK                   0.0              0.0  ...              0.0   \n",
       "KEAP1                 0.0              0.0  ...              0.0   \n",
       "MYD88                 0.0              0.0  ...              0.0   \n",
       "TBX3                  0.0              0.0  ...              0.0   \n",
       "\n",
       "gene_ids  ENSG00000196664  ENSG00000228933  ENSG00000224294  ENSG00000122824  \\\n",
       "DIS3                  0.0              0.0              0.0              0.0   \n",
       "MECOM                 0.0              0.0              0.0              0.0   \n",
       "ERBB2                 0.0              0.0              0.0              0.0   \n",
       "CHD7                  0.0              0.0              0.0              0.0   \n",
       "HCK                   0.0              0.0              0.0              0.0   \n",
       "KEAP1                 0.0              0.0              0.0              0.0   \n",
       "MYD88                 0.0              0.0              0.0              0.0   \n",
       "TBX3                  0.0              0.0              0.0              0.0   \n",
       "\n",
       "gene_ids  ENSG00000158639  ENSG00000269437  ENSG00000204025  ENSG00000165509  \\\n",
       "DIS3                  0.0              0.0              0.0              0.0   \n",
       "MECOM                 0.0              0.0              0.0              0.0   \n",
       "ERBB2                 0.0              0.0              0.0              0.0   \n",
       "CHD7                  0.0              0.0              0.0              0.0   \n",
       "HCK                   0.0              0.0              0.0              0.0   \n",
       "KEAP1                 0.0              0.0              0.0              0.0   \n",
       "MYD88                 0.0              0.0              0.0              0.0   \n",
       "TBX3                  0.0              0.0              0.0              0.0   \n",
       "\n",
       "gene_ids  ENSG00000126895  \n",
       "DIS3                  0.0  \n",
       "MECOM                 0.0  \n",
       "ERBB2                 0.0  \n",
       "CHD7                  0.0  \n",
       "HCK                   0.0  \n",
       "KEAP1                 0.0  \n",
       "MYD88                 0.0  \n",
       "TBX3                  0.0  \n",
       "\n",
       "[8 rows x 25691 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_template_df = icv.pp.create_cnv_template(adata, CNV_df)\n",
    "cnv_template_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eed6801-c666-438e-a20e-7193a71e4c3e",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Step 3: Simulate CNVs in an AnnData Object\n",
    "This step modifies the **gene expression matrix** in **adata** to simulate CNV effects.  \n",
    "The CNVs are applied to selected **subclones**, scaling expression values accordingly.\n",
    "\n",
    "**Process Overview**\n",
    "- **Initialize** CNV simulation layers in `adata.layers`.\n",
    "- **Assign subclone labels** to cells based on predefined groups.\n",
    "- **Modify expression values**:\n",
    "  - **Amplifications (gains)** → Increase gene expression.\n",
    "  - **Deletions (losses)** → Decrease gene expression.\n",
    "\n",
    "**Returns:**\n",
    "- Modified **AnnData object** (`adata`) with CNV-simulated expression values stored in:  \n",
    "  - `adata.layers['CNV_simulated']` → Scaled expression matrix.\n",
    "  - `adata.layers['CNV_GT']` → CNV ground truth matrix.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3629d27b-cf85-4ba0-871b-f511cdf6aeb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "subclone_dict = {\n",
    "    'N': [], # Normal cells without CNV simulations\n",
    "    'A': ['DIS3', 'MECOM', 'ERBB2', ], # Original genetic subclone\n",
    "    'B': ['DIS3', 'MECOM', 'ERBB2', 'CHD7', 'HCK', 'KEAP1'], \n",
    "    'C': ['DIS3', 'MECOM', 'ERBB2', 'CHD7', 'MYD88', 'TBX3'],\n",
    "} # 'DIS3','MECOM','ERBB2','CHD7','HCK','KEAP1','MYD88','TBX3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "de71bef0-7815-4214-81d1-8f2d5b4a8d1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers['counts'] = adata.layers['raw'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "07a98401-d9ad-4396-b6ca-a60d25eee42b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1268 × 25691\n",
       "    obs: 'organism_ontology_term_id', 'tissue_ontology_term_id', 'tissue_type', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'suspension_type', 'model_id', 'sample_id', 'Phase', 'level_1', 'level_2', 'level_3', 'CountUMIs', 'CountGenes', 'X.Mitochondrial', 'NoveltyScore', 'nCount_SCT', 'nFeature_SCT', 'orig.ident', 'is_primary_data', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'cnv_leiden', 'simulated_subclone'\n",
       "    var: 'gene_symbols', 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'feature_type', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'chromosome', 'start', 'end'\n",
       "    uns: 'cell_type_colors', 'citation', 'cnv', 'cnv_leiden', 'cnv_leiden_colors', 'cnv_neighbors', 'dendrogram_cnv_leiden', 'log1p', 'pca', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_cnv', 'X_cnv_pca', 'X_pca', 'X_ref.umap', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'logcounts', 'raw', 'counts', 'CNV_simulated', 'CNV_GT'\n",
       "    obsp: 'cnv_neighbors_connectivities', 'cnv_neighbors_distances'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "icv.pp.simulate_cnvs(adata, cnv_template_df, subclone_dict, cell_type_reference='cell_type', cell_type_cnv='secretory cell', alpha=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c5ad5b4b-4b49-49ce-9a8b-7e6801cc7f88",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"lung_organoids_cnvclust_simulatedCNVs_310125_simulation3_simulationv2_rho6.h5ad\", compression = 'gzip')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_env",
   "language": "python",
   "name": "scanpy_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
